docs/design-documents/drivers/serial/serial.md

```diff
@@ -0,0 +1,317 @@
+    1. [Deprecations](#deprecations).
+    1. [References](#references).
+
+
+### Revision history
+
+1.0 - Initial version - Evelyne Donnaes - 04/10/2019
+
+# Introduction
+
+### Overview and background
+
+Mbed OS contains multiple serial classes and multiple ways of printing information to the console. Serial classes have evolved over the years to add new functionality to existing classes and compensate for limitations. The result is that there are now multiple variants, not always well documented; therefore it is not clear which class should be used for what purpose. Furthermore, some of the classes pull substantial dependencies whose benefits are not very clear, for example standard library like objects and abstractions, and parts of the retarget layer.
+
+This document collects proposals made by various contributors (see [references](#references)) and presents a simplified serial class hierarchy. The aim is to offer two public classes, one for unbuffered I/O and the other for buffered I/O, with a clear definition of which one should be used when. All other serial classes are deprecated and should be removed at a later stage. In addition, the serial classes dependencies will be optimised to only include components that provide valuable functionality.
+
+#### Serial classes in Mbed OS 5.14
+
+**Serial**
+
+`Serial` provides unbuffered I/O and is using blocking HAL calls. It only works for simple printing. 
+
+Its main limitations are:
+- It uses a mutex lock so it cannot be used from interrupts. 
+- It wastes a lot of CPU time spinning waiting for serial output and it cannot be used reliably for input because it needs interrupt-speed response on read calls but its high-level API cannot be used from interrupts.
+- It lacks buffering so it cannot be used reliably for input and output without flow control.
+- It pulls in the C library stdio system because it uses `Stream`.
+- Although it is a `FileHandle`, its implementation is blocking only and has incorrect read semantics.
+
+**UARTSerial**
+
+`UARTSerial` provides buffered I/O. It can be used reliably for input from non-interrupt context and to avoid excess spinning waiting for transmission buffer space. It also does better blocking and correct POSIX-like `FileHandle` semantics. 
+
+It does not use `Stream` so it has the advantage of not pulling in the C stdio library. The drawback is that it doesn't include built-in printf methods, see [Usage scenarios and examples](#usage-scenarios-and-examples) for an example on how to print to a serial port.
+
+**RawSerial**
+
+`RawSerial` provides unbuffered I/O. It has no locks so it is safe to use in interrupts provided that it is the only instance that is using that serial port. Unlike `Serial` and `UARTSerial`, it is not a `FileHandle`.
+
+### Requirements and assumptions
+
+None
+
+# System architecture and high-level design
+
+Applications will have a choice between two public serial classes: 
+- `BufferedSerial` for a normal application
```

Comment:
I will remove.

Reply:
You still need to say something about why you'd use one or the other. The point is true that `BufferedSerial` should be the default choice, except for specialised cases.

`RawSerial` is something that doesn't even exist in any real operating system - you can't do it from application space at all, and you could only do it from inside the kernel by bypassing the normal drivers. You shouldn't be going near it unless you really know what you're doing.

(And the fact that the console defaults to `DirectSerial` is a pretty horrible legacy thing - should we finally fix that for Mbed OS 6?).

I can see valid reasons to use `RawSerial`, but they're right at the pointy end of absolutely maximising performance and absolutely minimising memory.

Reply:
We are removing `DirectSerial` now so it will come in 5.15 timeframe, which is a good thing!
You are right; I will rephrase to say that BufferedSerial should be the default choice.

Reply:
I was meaning the console defaulting to buffered serial in 6, rather than whatever the minimal/unbuffered version is called.

Reply:
`DirectSerial` is the default choice in the re-target code. My understand is that `MinimalConsole` should be the default choice in the re-target code.
@evedon Why should `BufferedSerial` be the default choice? I thought the default configuration of Mbed OS should be as small as possible.

Reply:
If an application uses `UnbufferedSerial` then it needs to understand the internals of serial in order to get expected I/O behaviour.

Reply:
Was [that](https://github.com/ARMmbed/mbed-os/pull/11634#discussion_r341530084) to answer my question?

Reply:
Unbuffered I/O involves reading or writing data one element at a time. The application needs to understand who has access to the serial port so that the writing or reading of characters is as expected. If you are writing from a critical section then you can expect a buffer of characters to be written in sequence; if you are not in a critical section then an unbuffered write could be interrupted if you don't have sole access to the serial port.

Using unbuffered I/O requires more knowledge so it should not be the default case. As written in the design `UnbufferedSerial` is provided for specialised cases like interrupt driven applications or applications that need to control access to the serial port.

We aim to provide a small memory footprint but the goal of the default configuration & API is easy of use and simplicity.

Reply:
My question was about the re-target code. Which class should the re-target code use by default?

Reply:
Sorry I got confused between the two different points. I was answering why BufferedSerial should be the default choice for an application.

The new internal MinimalConsole is the implementation when `stdio-minimal-console-only` is true. When it is false then we have:

```
static FileHandle *default_console()
{
#if MBED_CONF_TARGET_CONSOLE_UART && DEVICE_SERIAL
#  if MBED_CONF_PLATFORM_STDIO_BUFFERED_SERIAL
    static UARTSerial console(STDIO_UART_TX, STDIO_UART_RX, MBED_CONF_PLATFORM_STDIO_BAUD_RATE);
#   if   CONSOLE_FLOWCONTROL == CONSOLE_FLOWCONTROL_RTS
    console.set_flow_control(SerialBase::RTS, STDIO_UART_RTS, NC);
#   elif CONSOLE_FLOWCONTROL == CONSOLE_FLOWCONTROL_CTS
    console.set_flow_control(SerialBase::CTS, NC, STDIO_UART_CTS);
#   elif CONSOLE_FLOWCONTROL == CONSOLE_FLOWCONTROL_RTSCTS
    console.set_flow_control(SerialBase::RTSCTS, STDIO_UART_RTS, STDIO_UART_CTS);
#   endif
#  else
    static DirectSerial console(STDIO_UART_TX, STDIO_UART_RX, MBED_CONF_PLATFORM_STDIO_BAUD_RATE);
#  endif
#else // MBED_CONF_TARGET_CONSOLE_UART && DEVICE_SERIAL
    static Sink console;
#endif
    return &console;
}
```
If `MBED_CONF_PLATFORM_STDIO_BUFFERED_SERIAL` is true then BufferedSerial (UARTSerial) is selected. Otherwise we will use `UnbufferedSerial` instead of `DirectSerial`.

So, to answer your question, `platform.stdio-minimal-console-only` ? ("true" = MinimalConsole, "false" = UnbufferedSerial or BufferedSerial depending on `stdio-buffered-serial` is default)

Reply:
That actually did not answer my question 😀.

I have concluded that the default value of `platform.stdio-minimal-console-only` should be `false` as existing apps and tests rely on the APIs that are removed when it is set to `true`. Users have to enable `MinimalConsole` usage in the re-target code as it will be using `UnbufferedSerial\BufferedSerial` by default.

Reply:
> I thought the default configuration of Mbed OS should be as small as possible.

The general current pattern, that I don't think we're changing, is that full functionality is the default; so all the "minimal"s and "bare metals" are off by default.

We only configure "small" where it doesn't impact general functionality. Basically "optimise for space rather than speed" (eg minimise CRC tables), "or leave out unused stuff where possible"; but not "make features unavailable to save space", without explicit consent.

The logic of the existence and documentation of `BufferedSerial` suggests that it should be the default. The only two reasons I know of for it not to be are:

1. reducing memory footprint -- which would fall under the "minimal" default we just said we don't do
1. backwards compatibility with Mbed OS 2/3/5.

The backwards compatibility issues I'm aware of are:

1. Old-school attempts to separately declare and manipulate `Serial pc(USBTX, USBRX)` will likely fail if the console is using `BufferedSerial` - you get away with it with `UnbufferedSerial`/`DirectSerial`. This can hit either simple apps that can be simply changed to use `printf`, or complex ones that juggle console output with their own IRQ input. The latter would need re-engineering, or would have to manually set buffered console to false.
1. Default power consuption is high because reception IRQ is enabled; anyone wanting to enter deep sleep needs to do `disable_input`. (Maybe there could be a JSON setting to have the retarget do that for you?)